Track tablet streaming under global sessions to prevent side-effects of failed streaming

tgrabiec opened the pull request:
Tablet streaming involves asynchronous RPCs to other replicas which transfer writes. We want side-effects from streaming only within the migration stage in which the streaming was started. This is currently not guaranteed on failure. When streaming master fails (e.g. due to RPC failing), it can be that some streaming work is still alive somewhere (e.g. RPC on wire) and will have side-effects at some point later.

This PR implements tracking of all operations involved in streaming which may have side-effects, which allows the topology change coordinator to fence them and wait for them to complete if they were already admitted.

The tracking and fencing is implemented by using global "sessions", created for streaming of a single tablet. Session is globally identified by UUID. The identifier is assigned by the topology change coordinator, and stored in system.tablets. Sessions are created and closed based on group0 state (tablet metadata) by the barrier command sent to each replica, which we already do on transitions between stages. Also, each barrier waits for sessions which have been closed to be drained.

The barrier is blocked only if there is some session with work which was left behind by unsuccessful streaming. In which case it should not be blocked for long, because streaming process checks often if the guard was left behind and stops if it was.
 
This mechanism of tracking is fault-tolerant: session id is stored in group0, so coordinator can make progress on failover. The barriers guarantee that session exists on all replicas, and that it will be closed on all replicas.

kostja commented on service/storage_service.cc:
```diff
@@ -5948,6 +5965,22 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                        for (auto&& [tid, trinfo] : tmap.transitions()) {
+                            if (trinfo.session_id) {
+                                if (trinfo.stage == locator::tablet_transition_stage::write_both_read_new) {
+                                    close_session(session_id(trinfo.session_id));
```
I am trying to grasp how this is going to play out upon a coordinator change. On one hand session id is part of tablet metadata, so it's going to be loaded by the fresh coordinator, on the other, you don't wait for the new session id to be persisted in Raft before you close the previous session -  both events are happening in the background...

Could you formualte what you're trying to achieve in terms of invariants, so that one could try to validate them lookign at this code?

kostja replied:
Another concern is that going forward it would be rather difficult to ensure that these invariants are not broken:  in theory we could have some test coverage (a good idea anyway), but ideally, at least on the level of session tracking, we should assert them - e.g. that the session ids are monotonic, and that there are no more than two sessions against the same tablet (closed, established) at the same time.

tgrabiec replied:
There are two pieces here.

One goal here is to make sure that local set of open sessions mirrors sessions stored in group0. It is group0 state which determines whether a given session is open or closed. If session_id is set in tablet metadata, it means we want this session to be open. Ideally, we should close all sessions which don't have entries in tablet metadata. Maybe that would be more robust against future changes. Here, I do something simpler but a bit more hacky, which is to close them in write_both_read_new, where we know we want the session to be closed but still keep it in tablet info so that this path can pick up the session_id for closing. We know every replica will observe this because the coordinator executes a barrier in write_both_read_new. An alternative approach would be clear the session immediately when entering write_both_read_new and here close all sessions which are not in group0 state. 

Another piece is that the token metadata barrier executed by the coordinator ensures that each replica sees the group0 state, so will also reflect intended state of sessions, and guarantees that closed sessions are drained.

The highest invariant (0) we want to preserve is that there are no side-effects from operations started in the streaming stage once the coordinator moved to the next stage. In this implementation, this is ensured by successful execution of the step `tablet_transition_stage::write_both_read_old`, which executes a barrier, which will close and drain sessions used for the streaming stage.

kostja replied:
During our calls you were also mentioning in-flight requests. Session ids are only checked on rpc entry, and in-flight requests may be stuck in the reactor. How is this approach addressing this problem?

kostja replied:
Another concern is that session map is independent on tablet map, at least in C++ code. The code assumes that a tablet can not disappear from system.tablets without its session being closed. But is this really the case? There is no strong reference in the code between the two, in theory we could just forget the UUID of the session forever if the tablet is deleted without the respective session being destroyed. This, of course, shouldn't happen, well, at least in theory..

kostja replied:
> During our calls you were also mentioning in-flight requests. Session ids are only checked on rpc entry, and in-flight requests may be stuck in the reactor. How is this approach addressing this problem?

OK, this is actually clear, since they will be holding the gate and you drain the gate in close_sessions(), we'll wait for them... There is an unfortunate similarity of this approach with locks however. We don't know what resource the write will be holding and in turn waiting on but yet we assume that draining will succeed. As we get too many such subsystems in place, we will end up waiting on draining a session which is waiting on draining a fence or through workload prioritization get into some priority inversion and wait for an a completely unrelated semaphore or a gate.

tgrabiec replied:
> OK, this is actually clear, since they will be holding the gate and you drain the gate in close_sessions(), we'll wait for them... There is an unfortunate similarity of this approach with locks however. We don't know what resource the write will be holding and in turn waiting on but yet we assume that draining will succeed. As we get too many such fence/wait/drain subsystems in place, we will end up waiting on draining a session which is waiting on draining a fence or through workload prioritization get into some priority inversion and wait for an a completely unrelated semaphore or a gate.

Yes. What makes this problem less severe is the fact that this waiting only happens when there is a failed streaming, which is not a typical situation. When streaming is successful, there should be no guards alive for the closed session.

tgrabiec replied:
> Another concern is that session map is independent on tablet map, at least in C++ code. The code assumes that a tablet can not disappear from system.tablets without its session being closed. But is this really the case? There is no strong reference in the code between the two, in theory we could just forget the UUID of the session forever if the tablet is deleted without the respective session being destroyed. This, of course, shouldn't happen, well, at least in theory..

Yes, this code needs to be changed to be more robust in closing sessions.

kostja commented on service/storage_service.cc:
```diff
@@ -5948,6 +5965,22 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                        }
+                    }
+
+                    co_await drain_closing_sessions();
```
How are we protected against drain_closing_sessions being executed concurrently and/or out of order? We must assume that two coordinators can be acting concurrently against every node.

tgrabiec replied:
The intended post-condition of drain_closing_sessions() is that sessions which were closed prior to the call are drained after the call. 

Out of order execution is fine. The per-session close status and drain status is monotonic. Once session is closed it stays closed, and once it's drained it stays drained. Session stays around after it's closed and until drained, so later calls to drain_closing_sessions() will see sessions which are closed but not drained and call drain() on them. An absent session is both closed and drained. You can call session::drain() concurrently, and get the same post-condition (that it is drained).

I think there is a bug in drain_closing_sessions() where the iterator can be invalidated on concurrent call, I need to fix it by serializing the drain internally.

kostja replied:
What you describe is safe, but apparently not live. E.g. if there is an outdated coordinator running a request concurrently, or a leader change storm the effect will be just as gleb described - constant closing of sessions and no streaming progress made.

tgrabiec replied:
I don't see why we would have constant closing of sessions and no liveness. As long as we're in the "streaming" stage in group0, session remains open. Session is only closed after it is nulled in group0. Coordinator failover doesn't close it. The new coordinator will retry streaming under the same session.

If the streaming process is still alive (`storage_service::_tablet_ops` keeps a registry), retry will not start a new one but join with the currently running streaming.

What's the scenario with outdated leader running a request concurrently? What kind of request? Executing a barrier should not disrupt sessions, because barrier itself only makes the replica reflect what is in group0, and what is in group0 is what the current coordinator intends. If the session is open in group0, barrier from the old coordinator will not close it.

kostja replied:
> I don't see why we would have constant closing of sessions and no liveness. As long as we're in the "streaming" stage in group0, session remains open. Session is only closed after it is nulled in group0. Coordinator failover doesn't close it. The new coordinator will retry streaming under the same session.
> 
> If the streaming process is still alive (`storage_service::_tablet_ops` keeps a registry), retry will not start a new one but join with the currently running streaming.
> 
> What's the scenario with outdated leader running a request concurrently? What kind of request? Executing a barrier should not disrupt sessions, because barrier itself only makes the replica reflect what is in group0, and what is in group0 is what the current coordinator intends. If the session is open in group0, barrier from the old coordinator will not close it.

I assume the outdated leader will invoke the very same RPC. In some cases it could also be the previous or the next RPC of the topology change. You discussion about safety/liveness above pertains to a concrete case you have in mind - but you add a generic instrument, and at least by its definition it's not standing for a single case, it's a general purpose mechanism to manage streaming. So I am trying to define the general safety guidelines for using it, not just for the case of the current tablet state machine.
For example, I can imagine that under some circumstances a deposed leader may decide to abort streaming and thus close the sessions. In this case it will perhaps be able to close them because RPC is also guarded by term. But it indicates that the session framework itself is not conforming to the open/closed principle, i.e. it relies on an  assumption that lays outside the session API.

kostja replied:
Can out of order management of the session map happen? E.g. imagine there is a coordinator A that requests creation of session S1 via a state machine command. Then there is a coordinator change and S1 is closed. We search the session list and don't find S1 because it was not added there yet. Looks like  it's best to avoid any interleaved operations with the session map, but it's not yet clear how and whether this in itself will not lead to any deadlocks. I'd still say that if session ids are switched to timeuuids, and the erasure from the map is done based on "delete everything with timestamp < given timeuuid", rather than delete individual sessions by id, the whole system will be more robust (although perhaps not as flexible).

tgrabiec replied:
> Can out of order management of the session map happen? E.g. imagine there is a coordinator A that requests creation of session S1 via a state machine command. Then there is a coordinator change and S1 is closed. We search the session list and don't find S1 because it was not added there yet. Looks like it's best to avoid any interleaved operations with the session map, but it's not yet clear how and whether this in itself will not lead to any deadlocks. I'd still say that if session ids are switched to timeuuids, and the erasure from the map is done based on "delete everything with timestamp < given timeuuid", rather than delete individual sessions by id, the whole system will be more robust (although perhaps not as flexible).

Such problems cannot happen. Even though session management is invoked from the RPC, actual information does not come from the RPC, it comes from local group0 state machine. The meaning of RPC is "make sure sessions reflect group0". The order of such RPCs cannot affect the end result. So there is actually no concurrency regarding session changes. Local group0 follows the global group0 order, and local state of sessions follows the local group0 state, so it also follows the global group0 order of events.

Because group0 determines the state of sessions, we don't need session ids to be ordered.

avikivity commented on service/storage_service.cc:
```diff
@@ -7168,5 +7170,52 @@ future<> storage_service::node_ops_abort_thread() {
+    }
+}
+
+void close_session(session_id id) noexcept {
```
close() implies that after the call returns, we're done with the session (like close(2) or input_stream::close(). We need a name that suggests something else. Maybe start_drain_session().

tgrabiec replied:
Changed to start_closing().

avikivity commented on dht/range_streamer.hh:
```diff
@@ -78,7 +79,7 @@ public:
     range_streamer(distributed<replica::database>& db, sharded<streaming::stream_manager>& sm, const token_metadata_ptr tmptr, abort_source& abort_source, std::unordered_set<token> tokens,
             inet_address address, locator::endpoint_dc_rack dr, sstring description, streaming::stream_reason reason,
-            std::vector<sstring> tables = {})
+            std::vector<sstring> tables = {}, service::frozen_topology_guard topo_guard = {})
```
Should be non-optional to avoid untracked sessions by accident.

tgrabiec replied:
Ok

avikivity commented on service/storage_service.cc:
```diff
@@ -1611,7 +1612,15 @@ class topology_coordinator {
                 case locator::tablet_transition_stage::allow_write_both_read_old:
-                    transition_to_with_barrier(locator::tablet_transition_stage::write_both_read_old);
+                    if (do_barrier()) {
+                        updates.emplace_back(
```
Why not extend transition_to_with_barrier to accept a session?

avikivity commented on locator/tablets.hh:
```diff
@@ -161,10 +161,12 @@ struct tablet_transition_info {
     tablet_transition_stage stage;
     tablet_replica_set next;
     tablet_replica pending_replica; // Optimization (next - tablet_info::replicas)
+    utils::UUID session_id;
```
should be some tagged_uuid

tgrabiec replied:
Fixed

avikivity commented on service/session.hh:
```diff
@@ -0,0 +1,85 @@
+
+namespace service {
+
+using session_id = utils::UUID;
```
tagged_uuid

tgrabiec replied:
Fixed.

avikivity commented on replica/tablets.cc:
```diff
@@ -39,6 +39,7 @@ schema_ptr make_tablets_schema() {
             .with_column("replicas", replica_set_type)
             .with_column("new_replicas", replica_set_type)
             .with_column("stage", utf8_type)
+            .with_column("session", uuid_type)
```
Should it be list\<session> or something? So we can migrate tablets in different datacenters independently one day. Or even migrate multiple replicas in the same datacenter.

tgrabiec replied:
> Should it be list<session> or something? So we can migrate tablets in different datacenters independently one day. Or even migrate multiple replicas in the same datacenter.

Supporting more than one migration would require bigger changes to the schema. We cannot have one "stage" per tablet like we currently do. If you share the stage, you can as well share the session. If we're thinking about supporting this seriously, we should redesign the schema.

avikivity replied:
ok. since it's an internal table, we can perform internal migrations like adding a new column, backfilling it with data, and deleting the old column.

avikivity commented:
Looks reasonable. I can't review the raft integration like @kostja but the principle seems fine.

Let's have an API call (/debug) to retrieve sessions, we can later add more and more stuff to it so we don't need a core dump to track these things.

kostja commented on service/session.hh:
```diff
@@ -0,0 +1,85 @@
+
+using session_id = utils::UUID;
+
+/// Session is used to track execution of work related to some greater task, identified by session_id.
```
I think this comment needs to explain why session exists in addition to fencing, and how they operate on a different level. I also wonder whether it is a scylla concept or a lower level concept like seastar's. Essentially the way it is implemented here a sesison is a kind of an identifiable/findable gate. The whole persistence story exists outside session definition - which is, interestingly, suggests that sessions don't have to be persistent.

kostja replied:
For example, I would mention here that sessions are used to track long running RPCs based interactions which potentially involve *multiple* nodes and use many one-way RPCs to accomplish a task. That we purposefully want session identifiers to be independent from e.g. topology version, in order to keep some sessions open when a topology changes, and, vice versa, close some sessions without changing the cluster topology.

gleb-cloudius replied:
> which is, interestingly, suggests that sessions don't have to be persistent.

And frankly I do not see why this series bother to persist them. A session is always created in a state S1 and always removed in a state S2. S2 will never move to S3 unless the session is closed which means that session identifiers do not have to be persisted and can be chosen by each node randomly while preserving in memory mapping tablet_id->session_id. The code that looks for a session_id in tablet metadata will lookup it instead by calling `session_id get_streaming_session_id_for_tablet(tablet_id)`

tgrabiec replied:
> I think this comment needs to explain why session exists in addition to fencing, and how they operate on a different level. I also wonder whether it is a scylla concept or a lower level concept like seastar's. Essentially the way it is implemented here a sesison is a kind of an identifiable/findable gate.

This is a low level concept, so I think the description of the big picture doesn't belong in a comment here. Perhaps a docs/dev/topology-guard.md, and link to it.

tgrabiec replied:
> And frankly I do not see why this series bother to persist them. A session is always created in a state S1 and always removed in a state S2. S2 will never move to S3 unless the session is closed which means that session identifiers do not have to be persisted and can be chosen by each node randomly while preserving in memory mapping tablet_id->session_id. The code that looks for a session_id in tablet metadata will lookup it instead by calling `session_id get_streaming_session_id_for_tablet(tablet_id)`

Each node has to use the same session id so that you can pass the guard across nodes. If each node uses different session id, how do you detect if the guard is still valid?

gleb-cloudius replied:
Nodes do not just start streaming no? There is a node that contacts all other nodes and asks to stream data to it. This node chooses a random id for the session and pass it to all pears together with the streaming request. Isn't this what the patch "Propagate topology_guard to receivers " essentially doing?

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -7168,5 +7170,52 @@ future<> storage_service::node_ops_abort_thread() {
     __builtin_unreachable();
 }
 
+static thread_local std::unordered_map<session_id, std::unique_ptr<session>> sessions;
```
Since it has its own header may be we should put the implementation into a separate file as well. This one is already huge.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -3864,6 +3865,7 @@ future<> storage_service::drain_on_shutdown() {
 
 future<> storage_service::init_messaging_service_part(sharded<db::system_distributed_keyspace>& sys_dist_ks, bool raft_topology_change_enabled) {
     return container().invoke_on_all([&sys_dist_ks, raft_topology_change_enabled] (storage_service& local) {
+        create_session(service::session_id{}); // Default session.
```
What default session is for?

tgrabiec replied:
To be able to create a valid guard for streaming ops which don't use sessions created by topology code. If we have a default session, we can avoid special-casing for null guard.

tgrabiec commented on service/session.hh:
```diff
@@ -0,0 +1,85 @@
+session::guard enter_session(session_id);
+void create_session(session_id);
+void close_session(session_id) noexcept;
+future<> drain_closing_sessions();
```
I need them because the barrier should wait for all past sessions, not just those which are closed in the current invocation.

tgrabiec commented:
In V2:
   - Introduce session_manager
   - Move session management to session.cc
   - Update sessions from replica_to_all_cores() rather than the barrier. Sessions track group0 state, so this is more robust
   - Automatically close all sessions which are not in group0 rather than those which are in certain migration stage. More robust and easier to reason about correctness. Nulling a session in group0 will make sure it's closed. Introduced prune_sessions_except() to support this. Changed tablet migration to delete the session immediately after streaming.
   - Enable drain_closing_sessions() to be called concurrently (bug fix)
   - Improve documentation for session and topology_guard
   - Also use this mechanism to guard streaming for vnode-based topology changes on raft
   - Use tagged uuid for session id everywhere
   - Create default session in session_manager ctor rather than storage service initialization of messaging service
   - Rename session::close() to session::start_closing(), session::drain() to session::close()
   - Extracted get_mutation_builder() in handle_tablet_migration() to simplify code.

avikivity commented:
@gleb-cloudius please re-review

avikivity commented on configure.py:
```diff
@@ -1090,6 +1090,7 @@ def find_ninja():
                 'locator/util.cc',
                 'service/client_state.cc',
                 'service/storage_service.cc',
+                'service/session.cc',
```
CMakeLists.txt

asias replied:
I think the name session is too generic. Better to have a more specific name, e.g.,  global_work_sessions. The name is longer but it is more clear what the session is about.

asias replied:
In addition, we already have streaming/stream_session.cc.  It is too confusing to just call it session which will be mainly used in the stream_sessions.

avikivity commented on replica/tablets.cc:
```diff
@@ -104,13 +108,28 @@ tablet_mutation_builder::set_stage(dht::token last_token, locator::tablet_transi
 
+tablet_mutation_builder&
+tablet_mutation_builder::set_session(dht::token last_token, service::session_id session_id) {
+    _m.set_clustered_cell(get_ck(last_token), "session", data_value(session_id.uuid()), _ts);
```
We really need DML to work on these tables one day

gleb-cloudius replied:
Sometimes it would be an improvement, but other times an update is created dynamically, so having functions that create a mutation for each update and apply them atomically is easier then construct a string for the batch statement that you want to express.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -233,6 +233,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("unpublished_cdc_generations", set_type_impl::get_instance(cdc_generation_ts_id_type, true), column_kind::static_column)
             .with_column("global_topology_request", utf8_type, column_kind::static_column)
             .with_column("enabled_features", set_type_impl::get_instance(utf8_type, true), column_kind::static_column)
+            .with_column("session", uuid_type, column_kind::static_column)
```
Doesn't this need to be protected for upgrades?

avikivity commented:
Looks good. 

Missing:
 - test that would fail without this
 - docs

gleb-cloudius commented on service/session.hh:
```diff
@@ -0,0 +1,121 @@
+class session_manager {
+    std::unordered_map<session_id, std::unique_ptr<session>> sessions;
+    session::list_type closing_sessions;
+    seastar::semaphore session_drain_sem{1};
```
Those should be starting from underscore. Also I do not see where closing_sessions is populated.

gleb-cloudius commented on service/topology_guard.hh:
```diff
@@ -0,0 +1,72 @@
+/// which does. If the guard was invalidated while in transit, materialization will fail.
+/// Has an null state which represents no guard.
+/// topology_guard created out of null frozen_topology_guard is valid and never aborted.
+using frozen_topology_guard = session_id;
```
Nitpick but do we really need the type? Why not use session_id directly? Or, if you think we will need to extend it at some point, we can make session_id to be a struct with its own extensible IDL. IMO as it is `frozen_topology_guard ` provides nothing but obfuscation. It does not make the type extensible on wire.

gleb-cloudius replied:
Why not simple make `topology_guard` serializable?

asias replied:
From my experience with rpc, it is better to use a class for on wire objects (especially for the one that rpc verb returns) so we can extend later just in case. 

class frozen_topology_guard {
       session_id id;
}

tgrabiec replied:
> Why not simple make `topology_guard` serializable?

frozen_topology_guard is not just to be able to serialize the guard, it is another form of the guard which has slightly different semantics. You can copy it across shards and it doesn't hold the coordinator (yet).

gleb-cloudius commented on service/topology_guard.hh:
```diff
@@ -0,0 +1,72 @@
+public:
+    using frozen = session_id;
+
+    session_topology_guard(replica::table& t, frozen_topology_guard g)
```
We do not use 't' here. Will we have another guard which will use it?

tgrabiec replied:
If we used erm-based guards, which was an alternative solution based on per-tablet version, then we would need table&. I can drop it.

gleb-cloudius commented on dht/boot_strapper.hh:
```diff
@@ -52,7 +53,7 @@ public:
     }
 
-    future<> bootstrap(streaming::stream_reason reason, gms::gossiper& gossiper, inet_address replace_address = {});
+    future<> bootstrap(streaming::stream_reason reason, gms::gossiper& gossiper, service::frozen_topology_guard, inet_address replace_address = {});
```
Why local function gets frozen guard which suppose to be "in transit" representation?

tgrabiec replied:
Answered here: https://github.com/scylladb/scylladb/pull/15847#discussion_r1384195904

gleb-cloudius commented on dht/range_streamer.hh:
```diff
@@ -158,6 +161,7 @@ private:
     sstring _description;
     streaming::stream_reason _reason;
     std::vector<sstring> _tables;
+    service::frozen_topology_guard _topo_guard;
```
So the range streamer hold only a reference to the guard, but why? The guard needs to exists all the time the streaming is running, no? How do we guaranty it?

tgrabiec replied:
It doesn't have to exist all the time, only around section which has side-effects. If the coordinator moves on while the guard is not materialized, streaming will fail later when it tries to materialize the guard. It's even more efficient to not materialize early, because it won't block the coordinator unnecessarily.

gleb-cloudius commented on streaming/stream_session.cc:
```diff
@@ -111,24 +112,30 @@ void stream_manager::init_messaging_service_handler(abort_source& as) {
-    ms.register_stream_mutation_fragments([this, &as] (const rpc::client_info& cinfo, streaming::plan_id plan_id, table_schema_version schema_id, table_id cf_id, uint64_t estimated_partitions, rpc::optional<stream_reason> reason_opt, rpc::source<frozen_mutation_fragment, rpc::optional<stream_mutation_fragments_cmd>> source) {
+    ms.register_stream_mutation_fragments([this, &as] (const rpc::client_info& cinfo, streaming::plan_id plan_id, table_schema_version schema_id, table_id cf_id, uint64_t estimated_partitions,
+            rpc::optional<stream_reason> reason_opt,
+            rpc::optional<service::session_id> session,
```
So I wonder how is this different from the `plan_id`? As far as I know it is also a per "streaming session" uuid. Aren't we trying to invent here something that already exist? @asias

asias replied:
The plan_id is per stream_plan. We can use multiple stream_plans to stream data under a single "streaming session" to move data.

gleb-cloudius commented on streaming/stream_session.cc:
```diff
@@ -111,24 +112,30 @@ void stream_manager::init_messaging_service_handler(abort_source& as) {
+            rpc::source<frozen_mutation_fragment, rpc::optional<stream_mutation_fragments_cmd>> source) {
         auto from = netw::messaging_service::get_source(cinfo);
         auto reason = reason_opt ? *reason_opt: stream_reason::unspecified;
+        service::frozen_topology_guard topo_guard = session.value_or(service::session_id());
```
I thought that `frozen_topology_guard` is something to be used on wire, but the series uses almost everywhere, but on wire where simple uuid is transferred.

tgrabiec replied:
frozen_topology_guard is equivalent to uuid in this solution, so we also pass it on wire.

It's fine to pass it around the node too, since it's cheaper than the materialized guard. We need to materialize the full guard around operations which have side-effects, or if we wanted to check for abort early. In the plumbing layers it's unnecessary.

tgrabiec replied:
sessions are implementation detail of the guards, which I wanted to hide from topology guard users so that they are not coupled with a particular solution.

We need a distinction between frozen and regular guard, regardless of how you call it. "frozen" doesn't keep the topology coordinator on hold while it's alive, and it can be copied between shards and nodes by value, like frozen_mutation. Regular guard cannot be copied, and it will hold the topology coordinator.

gleb-cloudius commented on replica/tablets.cc:
```diff
@@ -195,8 +214,12 @@ future<tablet_metadata> read_tablet_metadata(cql3::query_processor& qp) {
+            service::session_id session_id;
+            if (row.has("session")) {
+                session_id = service::session_id(row.get_as<utils::UUID>("session"));
+            }
```
Since we set session_id only if the field is present it means that at other times it will default constructable, but  isn't this the same as default_session_id? Can it be dangerous?

tgrabiec replied:
Yes, it's better to initialize default_session_id to something else.

gleb-cloudius reviewed (commented):
I left some comments/questions.

asias commented on service/session.hh:
```diff
@@ -0,0 +1,121 @@
+    void create_session(session_id);
+
+    /// Calls start_closing() on all sessions except those in keep.
+    void prune_sessions_except(const std::unordered_set<session_id>& keep);
```
Why not just call it close_sessions_except()? "prune" makes it sound like we have more operations than closing.

tgrabiec replied:
I'm fine with close(), but then @avikivity will say that close() implies that we wait, but here we only start closing. So `initiate_close_of_sessions_except()`?

asias replied:
The new name sounds good.

asias commented on service/session.cc:
```diff
@@ -0,0 +1,68 @@
+}
+
+session_manager::session_manager() {
+    create_session(default_session_id);
```
Why do we need the default session? Is there alternatives to get rid of it?

tgrabiec replied:
It's to avoid special-casing for unguarded streaming. If we have default session, you can pass this session as the guard and materializing this guard will never fail because session is always there.

asias replied:
Perhaps we can call it unguarded_session_id instead of default_session_id. And we can have a service::get_unguarded_session_id() to get it. It is clearer than "default" session.

tgrabiec replied:
There are two levels of abstraction here, sessions and guards. The "unguarded" belongs to the guard. Default session is just a session which is always there. For the guard which is using it, we have `null_topology_guard`.

asias commented on service/session.cc:
```diff
@@ -0,0 +1,68 @@
+
+future<> session_manager::drain_closing_sessions() {
+    auto lock = co_await get_units(session_drain_sem, 1);
+    auto i = closing_sessions.begin();
```
Where do we insert the closing sessions  into closing_sessions? I do not see it.

tgrabiec replied:
Right, missed that.

asias commented on streaming/stream_plan.cc:
```diff
@@ -25,6 +25,7 @@ stream_plan& stream_plan::request_ranges(inet_address from, sstring keyspace, dh
     auto session = _coordinator->get_or_create_session(_mgr, from);
     session->add_stream_request(keyspace, std::move(ranges), std::move(column_families));
     session->set_reason(_reason);
+    session->set_topo_guard(_topo_guard);
```
Don't we need to call 

session->set_topo_guard(_topo_guard);


in stream_plan::transfer_ranges too?

asias commented on repair/row_level.cc:
```diff
@@ -495,8 +495,9 @@ void repair_writer_impl::create_writer(lw_shared_ptr<repair_writer> w) {
     }
     replica::table& t = _db.local().find_column_family(_schema->id());
     rlogger.debug("repair_writer: keyspace={}, table={}, estimated_partitions={}", w->schema()->ks_name(), w->schema()->cf_name(), w->get_estimated_partitions());
+    service::frozen_topology_guard topo_guard; // FIXME: propagate
```
By default,  we use repair to stream data for topology change. We need the topo_guard for repair  with higher priority than streaming.

tgrabiec replied:
Yes.

tgrabiec commented on service/topology_guard.hh:
```diff
@@ -0,0 +1,72 @@
+/// topology_guard created out of null frozen_topology_guard is valid and never aborted.
+using frozen_topology_guard = session_id;
+
+/// Represents a guard which is always valid and never aborted.
```
Will add one.

tgrabiec commented:
In V3:
- Make default_session_id be different than session_id{} for safety
- Ensure in code that default_session_id cannot be closed
- Prefix private fields of session_manager with _ 
- Actually close sessions
- Add unit test for session management
- Rename prune_sessions_except() to initiate_close_of_sessions_except()
- Introduce get_topology_session_manager()
- Added missing topology guard propagation to streaming code (not all paths were covered)
- Added test_tablets.py test case which reproduces consistency problem if sessions are disabled
- Added two commits from another PR to make the test case actually fail if sessions are reverted, otherwise progress would be held by erm:
- range_streamer, tablets: Do not keep token metadata around streaming
- range_streamer: Do not block topology change barriers around streaming
 - Added API to migrate a tablet, to make the test case possible
 - Added API to inject RPC connection drop, to make the test case possible
 - Added API to disable tablet load balancing
 - Added a small fix: tablets: Fail gracefully when migrating tablet has no pending replica
 - Added supporting error_injection changes:
- error_injection: Make is_enabled() public
- error_injection: Introduce poll_for_message()

Follow-ups:
- wire row-level repair to make all vnode-based topology changes safe
- add a similar test case of tablet repair once tablet repair is implemented

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:white_check_mark: - Unit Tests
#### Failed Tests (22/21917):
- [test_load_and_stream_decrease_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster[3_0_md])
- [test_load_and_stream_decrease_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster[3_0_md])
- [test_load_and_stream_decrease_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_mv_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_mv[3_0_md])
- [test_load_and_stream_decrease_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_mv_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_mv[3_0_md])
- [test_load_and_stream_decrease_cluster_with_index_view[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_index_view_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_index_view[3_0_md])
- [test_load_and_stream_decrease_cluster_with_index_view[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_index_view_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_index_view[3_0_md])
- [test_load_and_stream_2_columns_pk[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_2_columns_pk_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_2_columns_pk[3_0_md])
- [test_load_and_stream_2_columns_pk[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_2_columns_pk_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_2_columns_pk[3_0_md])
- [test_load_and_stream_increase_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_increase_cluster_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_increase_cluster[3_0_md])
- [test_load_and_stream_increase_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4934/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_increase_cluster_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_increase_cluster[3_0_md])
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4934/
- Duration: 2 hr 25 min
- Builder: spider5.cloudius-systems.com

avikivity commented:
@tgrabiec please also rebase

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:white_check_mark: - Unit Tests
#### Failed Tests (22/22023):
- [test_load_and_stream_decrease_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster[3_0_md])
- [test_load_and_stream_decrease_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster[3_0_md])
- [test_load_and_stream_decrease_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_mv_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_mv[3_0_md])
- [test_load_and_stream_decrease_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_mv_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_mv[3_0_md])
- [test_load_and_stream_decrease_cluster_with_index_view[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_index_view_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_index_view[3_0_md])
- [test_load_and_stream_decrease_cluster_with_index_view[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_decrease_cluster_with_index_view_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_index_view[3_0_md])
- [test_load_and_stream_2_columns_pk[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_2_columns_pk_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_2_columns_pk[3_0_md])
- [test_load_and_stream_2_columns_pk[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_2_columns_pk_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_2_columns_pk[3_0_md])
- [test_load_and_stream_increase_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_increase_cluster_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_increase_cluster[3_0_md])
- [test_load_and_stream_increase_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4986/testReport/junit/migration_test/TestLoadAndStream/Tests___Sanity_Tests___test_load_and_stream_increase_cluster_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_increase_cluster[3_0_md])
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4986/
- Duration: 3 hr 36 min
- Builder: i-0a13e4c78893117af

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5050/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6408/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5050/)
#### Failed Tests (22/22023):
- [test_load_and_stream_decrease_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_decrease_cluster_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster[3_0_md])
- [test_load_and_stream_decrease_cluster[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_decrease_cluster_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster[3_0_md])
- [test_load_and_stream_decrease_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_decrease_cluster_with_mv_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_mv[3_0_md])
- [test_load_and_stream_decrease_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_decrease_cluster_with_mv_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_mv[3_0_md])
- [test_load_and_stream_decrease_cluster_with_index_view[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_decrease_cluster_with_index_view_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_index_view[3_0_md])
- [test_load_and_stream_decrease_cluster_with_index_view[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_decrease_cluster_with_index_view_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_decrease_cluster_with_index_view[3_0_md])
- [test_load_and_stream_2_columns_pk[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_2_columns_pk_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_2_columns_pk[3_0_md])
- [test_load_and_stream_2_columns_pk[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_2_columns_pk_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_2_columns_pk[3_0_md])
- [test_load_and_stream_increase_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_increase_cluster_with_mv_3_0_md_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_increase_cluster_with_mv[3_0_md])
- [test_load_and_stream_increase_cluster_with_mv[3_0_md]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5050/testReport/junit/migration_test/TestLoadAndStream/Tests___dtest___test_load_and_stream_increase_cluster_with_mv_3_0_md__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_load_and_stream_increase_cluster_with_mv[3_0_md])
#### Build Details:
- Duration: 2 hr 24 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5103/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6469/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5103/)
#### Failed Tests (1/22015):
- [test_client_multipart_upload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5103/testReport/junit/boost/s3_test/test_client_multipart_upload) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_client_multipart_upload)
#### Build Details:
- Duration: 2 hr 28 min
- Builder: spider1.cloudius-systems.com

tgrabiec commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5103/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6469/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5103/)
> 
> #### Failed Tests (1/22015):
> * [test_client_multipart_upload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5103/testReport/junit/boost/s3_test/test_client_multipart_upload) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_client_multipart_upload)

```
[Exception] - storage_io_error: S3 request failed with (200 OK)
 == [File] - test/boost/s3_test.cc
 == [Line] - 151
```
Looks similar to https://github.com/scylladb/scylladb/issues/16133 \cc @xemul

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5104/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6474/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5104/)
#### Build Details:
- Duration: 2 hr 38 min
- Builder: spider4.cloudius-systems.com

xemul commented:
> ```
> [Exception] - storage_io_error: S3 request failed with (200 OK)
>  == [File] - test/boost/s3_test.cc
>  == [Line] - 151
> ```
> 
> Looks similar to #16133 \cc @xemul

Thanks. This means "request failed successfully" :exploding_head:
